### PR TITLE
Fix decompression bugs

### PR DESCRIPTION
**Pull request by bgilbert**
The xz decompressor didn't notice if compressed input data was truncated.  Also, the compression sniffer could miss a magic number that straddled a read buffer boundary.

### CODE REVIEW

File: src/io/peek.rs [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+// Read wrapper that allows peeking ahead in the stream without consuming
+// the peeked bytes.  BufRead.fill_buf() does not provide this, since it
+// only guarantees to return one byte.  For simplicity, we implement this
+// as a thin wrapper around BufReader.
+
+use bytes::{Buf, BytesMut};
+use std::io::{BufRead, BufReader, Read, Result, Seek, SeekFrom};
+
+pub struct PeekReader<R: Read> {
```
Comment (lucab) [inline src/io/peek.rs:23]: I did read through this Reader code and it seems sane, but I may have missed some edge cases. The testcases seems pretty extensive, so I mostly trust them for this logic.

**Review by lucab — APPROVED** [review]

File: src/io/peek.rs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::io::Cursor;
+
+    fn make_peek() -> PeekReader<Cursor<&'static [u8]>> {
+        // use BufReader capacity larger than input; we're not testing
+        // BufReader's buffering behavior
+        PeekReader::with_capacity(64, Cursor::new("abcdefghijklmnopqrstuvwxyz".as_bytes()))
```
[the remark this answers is not in the document]
Comment (cgwalters) [reply src/io/peek.rs:108]: (Also actually `Cursor::new` works on anything `AsRef<[u8]>` so you can just directly pass a `&str`)

[the remark this answers is not in the document]
Comment (bgilbert) [reply src/io/peek.rs:108]: Good call, fixed in https://github.com/coreos/coreos-installer/pull/920.

File: src/io/peek.rs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    pub fn with_capacity(capacity: usize, inner: R) -> Self {
+        Self {
+            source: BufReader::with_capacity(capacity, inner),
+            buf: BytesMut::new(),
+        }
+    }
+
+    /// Return the next amt bytes without consuming them.  May return fewer
+    /// bytes at EOF.
+    pub fn peek(&mut self, amt: usize) -> Result<&[u8]> {
```
[the remark this answers is not in the document]
Comment (bgilbert) [reply src/io/peek.rs:38]: Oh yeah, didn't think of that.  I think in this case the slice is a bit more usable, though.
